### DISCUSSION TITLE
PG logical replication disconnects - view updates are frozen

### DISCUSSION THREAD

**Discussion by Ten0** [MaterializeInc/materialize#29786] [Bug Reports] [answered]
Since updating to Postgres 16, some updates freeze (replication slots in `pg_replication_slots` are marked as `active:false, wal_status:lost`).

I was wondering whether this was a PG bug (they made lots of changes to logical replication in PG 16) or a Materialize bug, so I wanted to search in the issue tracker, but apparently all issues are gone now. (Also, the "bug reports" discussions are disabled.)

1. Does this particular bug ring a bell, and what would you expect its status to be?
2. How can we access the list of issues to track whether such an issue has already happened and been fixed? ([CONTRIBUTING.md mentions issues](https://github.com/MaterializeInc/materialize/blob/e47f5d2f5a1328ad6577e83c6a387550bca377ab/CONTRIBUTING.md?plain=1#L30-L31) but link is broken, and [Bug reports discussions categories is empty](https://github.com/MaterializeInc/materialize/discussions/categories/bug-reports?discussions_q=category%3A"Bug+Reports") (instead of having been migrated from existing issues).

Thanks,

**Comment by morsapaes** [comment]
Thanks for reporting this, @Ten0 — and sorry for the confusion! We're mid-migration and there are some loose ends to tie still while issues aren't publicly accessible. cc @chaas

We're testing against [multiple versions of PostgreSQL](https://github.com/MaterializeInc/materialize/blob/c0e1608fcbd832c36a87239f96e9b8bef83002ab/ci/release-qualification/pipeline.template.yml#L284) in CI (including 16.4), but haven't seen the specific error you mentioned before. What minor version are you seeing this in? We'll take a look internally, thanks again for reporting this!

**Reply by Ten0** [reply]
Hello!
We are using PG version 16.4 at the moment, and the issue started when we migrated from PG <16 to PG 16.3.
That happens to be at the same time as PG seemed to have made major changes to its logical replication according to the changelog.
The issue happens not immediately but rather randomly after a while (like a few hours of running).
Good to know that you haven't seen this issue before :+1: 
I was hoping it was a PG bug and it would fix itself with the next point release but seeing as that hasn't happened I'm dedicating more time to this now.

Checking the PG logs I can see the following:
```
ERROR: can no longer get changes from replication slot "materialize_..."
DETAIL: This slot has been invalidated because it exceeded the maximum reserved size.
```
Maybe the update has increased how much storage it actually requires, or maybe as a coincidence we just started using more, but maybe this is rather a symptom rather than the cause (it stopped reading -> started storing more -> died due to too much storage used).
For now I'll increase `max_slot_wal_keep_size` x4, reset the replication slots and see if that solves anything.

**Reply by Ten0** [reply]
So the logs I'm getting are a `terminating connection due to administrator command`, then 3 seconds later materialize retries connecting and `can no longer get changes from replication slot`, `DETAIL: This slot has been invalidated because it exceeded the maximum reserved size`.
This started happening when we migrated from pg 15 to pg 16, but since then we have increased our `max_slot_wal_keep_size` from 512M to 10G and the same error keeps happening, disconnecting after a while running, so it looks like it's not just PG 16 consuming slightly more...
Do you have an idea what would cause this, or if perhaps someone else has encountered the same issue?

![image](https://github.com/user-attachments/assets/afc0e65c-9bcc-42f8-b756-00466692a602)
![image](https://github.com/user-attachments/assets/07b4886a-fee3-46e1-886e-d841165e7847)
(I think the drops are when we re-create everything replication-slot/materialize-related to make it refresh)

**Reply by morsapaes** [reply]
What version of Materialize are you running?

**Reply by Ten0** [reply]
v0.26.6 (5ad960ec7) 😁

**Reply by morsapaes** [reply, marked as answer]
Ah. Sorry — we no longer support versions under v0.27 as of October 2022 (as stated in the [LTS documentation](https://materialize.com/docs/lts/release-notes/)). The PostgreSQL source has been refactored for the new architecture and hardened based on feedback from users running it in production against different PostgreSQL versions (and services), which explains why we haven't seen this issue with newer Materialize versions. If you'd like to keep using the binary version, I'm afraid you'll need to stick to PostgreSQL < v16.

**Reply by Ten0** [reply]
I'm trying to experiment on whether upgrading would solve this issue.
Trying to connect to a newer version works fine with `psql`. I'm trying to connect via datagrip following [this guide](https://materialize.com/docs/integrations/sql-clients/#datagrip), which used to work in `0.26`, but it is now unable to connect (maybe it takes the connection NOTICE message for an error):
![image](https://github.com/user-attachments/assets/33f7bdb3-3142-43dd-8328-d2f15c71b87c)

Since issues are gone, where can I report this?
Thanks,

EDIT: I can confirm that is indeed the issue: Specifying `options: --welcome_message=off` in DataGrip allows it to connect:
![image](https://github.com/user-attachments/assets/b9659165-6654-4302-b31a-80c23321c3c0)
So maybe only the guide needs a fix, although it would be more practical if upon recognizing DataGrip it would also turn of the welcome message...

 => Is there any way to turn the message off globally, by e.g. adding an env var or a config file in the docker?
I found `MZ_SYSTEM_PARAMETER_DEFAULT=welcome_message=off` but that doesn't change anything.

**Reply by morsapaes** [reply]
Ah, that's an interesting finding. I'll tweak the documentation to add the `welcome_message` tip! You can set the default value of `welcome_message` to `off` for the connecting role by running the following command (which will persist across sessions):

```sql
ALTER ROLE materialize SET welcome_message = off;
```

There's an internal tracking issue that proposes turning this off by default at the system level, or allowing to customise the value depending on the `application_name`, as you suggested. We explicitly turn this off in our dbt adapter, too — so it's definitely open for debate! I added this thread to the issue for consideration.
